### PR TITLE
fix(qwen3-moe): export grouped HF expert weights

### PR DESCRIPTION
**Pull request by akoumpa**
## What
Fix Qwen3-MoE HF checkpoint export/import so full-model expert weights use the grouped Transformers v5 key layout:

- `model.layers.*.mlp.experts.gate_up_proj`
- `model.layers.*.mlp.experts.down_proj`

This keeps `v4_compatible=True` on the legacy split per-expert export path.

## Why
The scheduled nemo-ci job `qwen3_moe_30b_te_deepep` failed checkpoint robustness because AutoModel reload from the consolidated checkpoint was exact, but vanilla HF reload diverged:

- Phase 3 AutoModel reload KL: `0.000000e+00`
- Phase 4 HF reload KL: `8.225249e-02 > 1e-2`

The installed HF Qwen3-MoE implementation expects grouped expert tensors, while the adapter was exporting split per-expert keys.

## Changelog
- Export Qwen3-MoE full-model expert tensors in grouped HF format by default.
- Import grouped HF expert tensors back into native AutoModel layout.
- Keep legacy split export behind `v4_compatible=True`.
- Make the focused Qwen3-MoE adapter unit tests CPU-runnable and cover grouped/legacy paths.

## Pre-checks
- `python -m pytest tests/unit_tests/models/qwen3_moe/test_qwen3_moe_state_dict_adapter.py -q`
- `python -m ruff check nemo_automodel/components/models/qwen3_moe/state_dict_adapter.py tests/unit_tests/models/qwen3_moe/test_qwen3_moe_state_dict_adapter.py`
- Tiny HF Qwen3-MoE grouped expert state-dict round trip through the adapter.
- DCO sign-off included.

### CODE REVIEW

**Comment by akoumpa** [comment]
nemo-ci validation passed for 6a2cb2e2f922fcf7fa5e1b4e1bbdb4cf9a84843f.

- Root pipeline 55613452: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55613452
- Generated LLM pipeline 55614109: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55614109
- qwen3_moe_30b_te_deepep job 346373697: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/346373697

**Comment by akoumpa** [comment]
nemo-ci validation passed for 3ef8fd50f4a5058fb9e0408545cf9eece5e15dd6 without modifying examples/llm_finetune/qwen/qwen3_moe_30b_te_deepep.yaml.

- Root pipeline 55640193: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55640193
- Child pipeline 55640356: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55640356
- Generated LLM pipeline 55640373: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55640373
- qwen3_moe_30b_te_deepep job 346581220: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/346581220

The qwen3_moe_30b_te_deepep recipe remains unchanged in the PR diff; the downstream robustness-only HF reload fixtures are applied by the CI resolver.

**Review by HuiyingLi — DISMISSED** [review]

**Comment by HuiyingLi** [comment]
Pushed cfa9532b: dropped the grouped-experts HF **export** change (adapter + checkpointing + their tests), keeping only the CI knobs that actually fix the divergence (`experts_implementation: eager` + relaxed `hf_kl_threshold`).

Rationale: transformers v5 already auto-converts the split per-expert layout → grouped `gate_up_proj`/`down_proj` at load time, so the grouped export is unnecessary and just makes the saved checkpoint less portable (split is the canonical layout older transformers / vLLM / tooling expect).

Verified on **transformers 5.5.0** (container `auto2604`): a split per-expert checkpoint loaded via `from_pretrained` gives **bit-identical** logits to the grouped checkpoint and to the reference model:

```
grouped-load vs ref          : max|d| = 0.000e+00
split-load   vs ref          : max|d| = 0.000e+00
split-load   vs grouped-load : max|d| = 0.000e+00
```

This also matches Phase-3 (AutoModel reload) KL = 0.0: the weights were always loading correctly regardless of format; the 8.2e-2 Phase-4 divergence was fused-`grouped_mm` bf16 numerics (router flips), which `experts_implementation: eager` addresses.

**Comment by HuiyingLi** [comment]
nemo-ci validation passed for cfa9532bb7d8564c958304eca3ffbb729e305ac1 (grouped-export change reverted; `experts_implementation: eager` + threshold kept).

- Root pipeline 57626571: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57626571
- Child pipeline 57626693: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57626693
- Generated LLM pipeline 57626702: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57626702
- qwen3_moe_30b_te_deepep job 360630775: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/360630775

The `qwen3_moe_30b_te_deepep` checkpoint-robustness recipe passes **without** the grouped HF expert export — confirming that change is unnecessary (transformers v5 auto-converts the split per-expert layout losslessly), and the divergence fix comes from `experts_implementation: eager` + the relaxed `hf_kl_threshold`.

File: tests/ci_tests/scripts/config_templates/ci_config.yaml [hunk 1]
```diff
@@ -77,6 +77,16 @@ conditional:
     apply:
       peft.use_triton: false
 
+  # Qwen3-MoE DeepEP's checkpoint correctness is covered by the strict
+  # Automodel reload phase.  The vanilla-HF reload phase compares against a
+  # different bf16 MoE backend, so keep that CI tolerance in the resolver
+  # rather than conditioning the public recipe on the downstream HF path.
+  - when_recipe_contains_all: [qwen3_moe_30b_te_deepep]
```
Comment (akoumpa) [inline tests/ci_tests/scripts/config_templates/ci_config.yaml]: @HuiyingLi is there a way to avoid config-specific changes to infra? Also, i see `checkpoint_robustness.experts_implementation: eager`, i imagine there might be other options, are those unsupported?

Thank you

Comment (HuiyingLi) [reply tests/ci_tests/scripts/config_templates/ci_config.yaml]: Thank! Fixed

Comment (HuiyingLi) [reply tests/ci_tests/scripts/config_templates/ci_config.yaml]: For the eager, it is not for training. It is only for forcing hf to load the vanilla splitted experst. bc otherwise there is numerical drift, and then we see the large kl.

**Review by HuiyingLi — APPROVED** [review]
